Create a log storage capsule

TPackard opened the pull request:
### Pull Request Overview

This pull request adds a log storage capsule that offers an interface for storing persistent data in flash. Logs are implemented on top of flash volumes that are statically allocated at compile time. Data can be written to logs in the form of entries, which can later be read back sequentially. Log data persists across device reboots, and logs automatically recover any log previously saved data from flash after booting. Logs can either be linear (denying writes when full) or circular (overwriting the oldest entries with the newest ones when the underlying flash volume is full).

### Testing Strategy

A test suite for the log storage capsule was created for the imix board. One main test (`boards/imix/src/storage_test.rs`) exercises a circular log across a wide variety of log operations and failure cases, checking to make sure that the log behaves as expected and handles errors gracefully. A second test (`boards/imix/src/linear_storage_test.rs`) tests a linear log to ensure that writes are properly denied when a linear log is full.

### TODO or Help Wanted

Let me know if I did anything weird/wrong that I should change

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

hudson-ayers commented on kernel/src/hil/flash.rs:
```diff
@@ -111,10 +111,18 @@ pub trait Flash {
+        &self,
+        page_number: usize,
+        buf: &'static mut Self::Page,
+    ) -> (ReturnCode, Option<&'static mut Self::Page>);
```
These seem like good changes to the HIL, but you only modified the sam4l implementation of the HIL, so know all other chips will fail to compile (which is why Travis is failing). I think it makes sense to break out the changes to the HIL / / chip-specific flash drivers / non-log capsules into its own PR, which can be reviewed and merged independently of the log capsules and tests for the log capsule.

TPackard commented:
Removed the flash hil changes that returned client buffers on failure, I'll submit these later in their own PR after making them work for other boards.

hudson-ayers commented on capsules/src/log_storage.rs:
```diff
@@ -0,0 +1,865 @@
+            offset += 1;
+        }
+
+        assert!(offset <= ENTRY_HEADER_SIZE);
```
given that ENTRY_HEADER_SIZE is defined as sizeof(usize), and this for loop iterates through the number of bytes in a usize, this assert statement seems superfluous

phil-levis commented on capsules/src/storage_interface.rs:
```diff
@@ -0,0 +1,75 @@
+}
+
+/// An interface for reading from log storage.
+pub trait LogRead {
```
This should have a method to set a LogReadClient.

phil-levis commented on capsules/src/storage_interface.rs:
```diff
@@ -0,0 +1,75 @@
+}
+
+/// An interface for writing to log storage.
+pub trait LogWrite {
```
This should have a method to set a LogWriteClient.

phil-levis commented:
>     1. Why separate the LogRead and LogWrite traits? It is difficult to imagine a scenario where the same struct does not implement both, or the purpose of a read-only or write-only log.

The motivation for finer-grained traits is for the callers, rather than the implementations. Just because a capsule can write to the log does not mean it should be able to read it. It also separates out client callbacks.

>     2. Separating `LogReadClient` and `LogWriteClient` makes more sense to me, as I could imagine a setup where most capsules are configured as Write-Only clients to the log, and then only a single error reporting capsule or something is a `LogReadClient`. But the way the LogStorage capsule is currently implemented makes it impossible for a client to register without being both a `LogReadClient` and a `LogWriteClient` anyway.

You want separate Read and Write traits, both for the calls and the Clients.


>     3. This isn't really a question about this particular PR, but I believe that currently flashing a new kernel does not erase all of flash (as currently flashing a new kernel leaves apps running on the board) -- but we would probably want flashing a new kernel to also clear this log, yes? Does that happen? Can we document this somewhere?

I think it should. The log (or rather, a kernel flash storage region) is modeled as a large zero-filled variable in the image. So it's big block of zeros in the image that's copied in. This also means the log does not persist across kernel updates, because the variable can move around. However, I think your point is valid, that we should state that this is an assumption.
 
>     5. This is less of a big deal, but this PR does add 12 calls to `unwrap()`, `expect()`, or `panic!()` to capsules, along with several places where passing incorrect parameters to functions could cause panics on array accesses. Tracing through these I didn't see any where I could immediately think of how to trigger them, but if the ultimate goal is to use this capsule as an in-kernel error reporting mechanism when panics occur elsewhere, it would be nice to be confident that this capsule itself should not panic.

Agreed.

TPackard commented:
> Why separate the LogRead and LogWrite traits? It is difficult to imagine a scenario where the same struct does not implement both, or the purpose of a read-only or write-only log.

The original thinking was to allow for read-only or write-only clients, although only allowing for read/write clients made that useless. I split the client setting function to make this work properly.

> Any particular reason for separating the traits in `storage_interface.rs` into their own file instead of including them in `log_storage.rs`? I think it is fine if you want to keep it that way, but think `log_interface.rs` would be a better filename.

Originally, there were plans to eventually create a couple different kinds of flash storage interfaces. I can rename `storage_interface.rs` to `log_interface.rs` until these other interfaces are actually created if that makes more sense.

> This is less of a big deal, but this PR does add 12 calls to `unwrap()`, `expect()`, or `panic!()` to capsules, along with several places where passing incorrect parameters to functions could cause panics on array accesses. Tracing through these I didn't see any where I could immediately think of how to trigger them, but if the ultimate goal is to use this capsule as an in-kernel error reporting mechanism when panics occur elsewhere, it would be nice to be confident that this capsule itself should not panic.

I wasn't quite sure what to do here, but there are a few instances where the log could fail if misconfigured/buggy and was not in a position to return an error to the client. I didn't want it to fail silently, so I added the panics/unwraps. Given that we don't ever want the log to cause a panic when used for error logging, is there a better alternative to just failing silently?

Additionally, there are certain errors that the log cannot recover from, such as losing the pagebuffer from its `TakeCell`. This should be impossible once the flash hil returns buffers in the event of an error, but what would it be appropriate to do if this did happen?

For now, I'll add an invalid state to the log that will deny all further operations in the event of an unrecoverable error. I left a couple of unwrap statements, but only for converting byte arrays to `usize`s. These shouldn't ever fail because the slices being converted are always `size_of::<usize>()` bytes long.

Also added comments to the code to clarify some of the points made above.

phil-levis commented on capsules/src/storage_interface.rs:
```diff
@@ -0,0 +1,73 @@
+use kernel::ReturnCode;
```
Should this file be in kernel::hil?

TPackard replied:
True, moving it into the hil

phil-levis commented on capsules/src/log_storage.rs:
```diff
@@ -0,0 +1,936 @@
+    error: Cell<ReturnCode>,
+}
+
+impl<'a, F: Flash + 'static, RC: LogReadClient, WC: LogWriteClient> LogStorage<'a, F, RC, WC> {
```
Does this need to be parameterized by the LogReadClient and LogWriteClient types, or could they be dyn?

TPackard replied:
Oh yeah, using dyn is much cleaner, thanks.

bradjc commented on capsules/src/log_storage.rs:
```diff
@@ -0,0 +1,936 @@
+//! Implements a log storage abstraction for storing persistent data in flash. Data entries can be
```
```suggestion
//! Implements a log storage abstraction for storing persistent data in flash.
//!
//! Data entries can be
```

TPackard commented:
I re-added panics in the event of unrecoverable errors (like not getting a buffer from a TakeCell that should always have a buffer) after discussing this with Phil and Hudson earlier in the week.


> I think it is a bit of missed opportunity that this uses a `hil::Flash` driver rather than `hil::NonvolatileStorage`, as boards can't use FRAM or other nonvolatile storage, but maybe there isn't a use case for not using the on-board flash.

I'm not super clear on what the difference between flash and nonvolatile storage in the hil is, can anybody clarify?

bradjc commented:
> > I think it is a bit of missed opportunity that this uses a `hil::Flash` driver rather than `hil::NonvolatileStorage`, as boards can't use FRAM or other nonvolatile storage, but maybe there isn't a use case for not using the on-board flash.
> 
> I'm not super clear on what the difference between flash and nonvolatile storage in the hil is, can anybody clarify?

Nonvolatile storage is just a bit of a more general interface that doesn't assume flash's page-based characteristics. So something like FRAM, which is byte read- and write-able, would have a driver that would implement `hil::NonvolatileStorage`, rather than `hil::Flash`. There is also a converter capsule that takes a flash interface and provides NonvolatileStorage.

However, it may be very problematic to ignore the realities of flash when implementing the log storage on top of flash. Perhaps it actually makes more sense to have two implementations of log storage, one that assumes flash pages and the other which doesn't.

bradjc commented on kernel/src/hil/storage_interface.rs:
```diff
@@ -0,0 +1,74 @@
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum StorageCookie {
+    SeekBeginning,
+    Cookie(usize),
```
What is the `usize`? Maybe the `Cookie { field: type }` version would be better.

TPackard replied:
Just a value to represent some entry in the log, how it does so is left to the implementer. Would it make more sense to give this value a name?

bradjc commented on kernel/src/hil/storage_interface.rs:
```diff
@@ -0,0 +1,74 @@
+use crate::returncode::ReturnCode;
+
+pub type StorageLen = usize;
+pub type OperationResult = Result<(), (ReturnCode, Option<&'static mut [u8]>)>;
```
I would say remove these. We don't do this elsewhere in Tock HILs (I don't think).

bradjc commented on kernel/src/hil/storage_interface.rs:
```diff
@@ -0,0 +1,74 @@
+//! Interfaces for storage devices.
```
Perhaps rename this file `log.rs`?

TPackard replied:
I believe that there is a plan to implement a couple other storage interfaces in the future. Is it alright to continue calling this `storage_interface.rs` under the assumption that they will be added to this file some time in the future, or should this file be renamed and the new storage interfaces go into their own files?

bradjc replied:
If they share elements (like counter/alarm/timer) then a single file makes sense (although it should still be called "storage.rs". If they don't, then it seems easier to just have them be separate.

We actually have examples both ways on this (i2c/spi/uart are not grouped, but ninedof/humidity/light/temp are). I'm not sure which is better, but from a git point of view it is usually easier to have different files.

bradjc commented on kernel/src/hil/storage_interface.rs:
```diff
@@ -0,0 +1,74 @@
+//! Interfaces for storage devices.
+
```
What isn't clear to me is the log interface record based, or just byte level, or something else? If I seek to position "2", does that mean the third thing written to the log? Or the third byte in the log?

TPackard replied:
The log interface is record based, although records are identified by the position of their first byte. Seeking to position "2" will seek to the 2nd byte under the assumption that a record starts there. The log can handle seeks to invalid positions, although it won't always immediately realize that a seek was invalid.

bradjc replied:
I see, thanks for the clarification. I read through https://github.com/tinyos/tinyos-main/blob/master/doc/txt/tep103.txt#L328, and it's not clear to me how users of the log should read from it. Are users of the log responsible for setting up some sort of delineation between log entries? The TEP doesn't go into much detail on this aspect (and even argues that the log interface is high level!), but I'm not sure if your implementation takes a different approach. I think a description of how something should use the log would be helpful.

TPackard replied:
The log internally keeps track of the size of entries and only reads back whole entries. It will read the entries back sequentially by default, and seeking allows for reading entries out of order. However, it assumes that the user will only seek to valid entry positions (retrieved through the `current_read_offset` and `current_append_offset` functions). I could also add some additional data to the entry headers in order to better verify that a valid position was seeked if that seems necessary.

bradjc replied:
Whatever we converge on should be documented here.

TPackard commented:
> > > I think it is a bit of missed opportunity that this uses a `hil::Flash` driver rather than `hil::NonvolatileStorage`, as boards can't use FRAM or other nonvolatile storage, but maybe there isn't a use case for not using the on-board flash.
> > 
> > 
> > I'm not super clear on what the difference between flash and nonvolatile storage in the hil is, can anybody clarify?
> 
> Nonvolatile storage is just a bit of a more general interface that doesn't assume flash's page-based characteristics. So something like FRAM, which is byte read- and write-able, would have a driver that would implement `hil::NonvolatileStorage`, rather than `hil::Flash`. There is also a converter capsule that takes a flash interface and provides NonvolatileStorage.
> 
> However, it may be very problematic to ignore the realities of flash when implementing the log storage on top of flash. Perhaps it actually makes more sense to have two implementations of log storage, one that assumes flash pages and the other which doesn't.

I see, thanks. I think I'll leave this as it is now and see if only working with flash interfaces becomes problematic.

bradjc commented:
I can see having a log interface that is more like a `/var/log/` file, where the kernel can write whatever it wants, and there is a convention that log entries end in `\n`, but the abstraction is really an append only log with a lot of flexibility for users. A log interface that has a much stricter notion of records and entries would also make sense. But I would expect the interface to be clear as to what the client can expect and how the client should use the log.

TPackard commented:
Tried to clarify how entries are represented via cookies and what constitutes a legal seek. Hopefully this helps, but let me know if anything is still unclear.

The log interface only allows for reads/writes/seeks of whole log entries, operations outside of entry boundaries are not allowed.

bradjc commented on kernel/src/hil/storage_interface.rs:
```diff
@@ -0,0 +1,79 @@
+/// particular position within the interface. How a `Cookie` internally represents a position is up
+/// to the implementer.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum StorageCookie {
```
It seems like storage cookie is both a public and private type. Since it has implementation-specific content, why not make it an associated type with the trait so there is no confusion that users shouldn't edit or create it?

TPackard replied:
I'm not quite sure how that would work since the storage cookie type is used by both the `LogRead` and `LogWrite` traits. I'd also like to require that the storage cookie be an enum with the `SeekBeginning` variant, and I'm not sure how to do that if it's an associated type. Any recommendations on how to go about this? I agree that it would be much nicer if users weren't able to create storage cookies on their own.

bradjc replied:
There is maybe a way to do that, since the flash hil has to do something similar? But, I don't think I have a good mental model for how the cookies should be used, and why the internal state of how to read the log needs to be exposed through the cookie type. If you could add a description of the intent of the cookie in the HIL that would be very helpful.

TPackard replied:
I don't currently see a good way to refactor the `StorageCookie` enum as an associated type, so I'm going to leave it as is for now.

bradjc commented on kernel/src/hil/storage_interface.rs:
```diff
@@ -0,0 +1,79 @@
+    fn current_read_cookie(&self) -> StorageCookie;
+
+    /// Seek to a new read position.
+    fn seek(&self, offset: StorageCookie) -> ReturnCode;
```
How do clients advance to the next record in the log?

TPackard replied:
The log automatically advances to the next record after every read. I realized that I forgot to update the documentation on the hil, so I'll fix that.

TPackard commented:
Refactored the naming to simply refer to this as the `Log` rather than `StorageInterface` as that seems to make sense to most people.

I also tried to improve the documentation and explanations some more.

ppannuto commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,88 @@
+    /// Set the client for appending from a log. The client will be called when writing operations complete.
+    fn set_append_client(&'a self, append_client: &'a dyn LogWriteClient);
+
+    /// Append an entry to the end of the log.
```
Is this re-entrant or must callers wait until `append_done` before calling again? [question motivated by `record"s"_lost` below]

TPackard replied:
Left up to the implementer, it is not re-entrant in the log capsule implementation.

https://github.com/tock/tock/blob/866d3f36241df00b2bc364a245da0b474467b010/capsules/src/log.rs#L692

ppannuto commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,88 @@
+        &self,
+        buffer: &'static mut [u8],
+        length: usize,
+        records_lost: bool,
```
A bit confused by the plural here, can there be multiple records lost?

Why are `records_lost` different from `error`? [Only some errors lose records?]

If records are "lost", are they actually still in the `buffer` returned here?

TPackard replied:
Yes, multiple records can be lost, and it is not an error. See below.

TPackard commented:
> With apologies for being way behind on Tock reviews, I looked over just the HIL changes here and have some questions. I think my biggest hangup is the "cookie" thing. I don't quite understand what this mechanism is trying to do, for a kernel logging mechanism, why do we need anything more than some stateless record delimiter?
> 
> Are cookies part of a log entry? In which case why is a log beginning cookie possible?
> 
> The log itself is a shared resource whose usage and fill rate is unpredictable to any one accessor of the log. Holding a cookie and assuming it will be present on next read feels bug prone?

Cookies are identifiers generated by the log and only used by the client to seek log entries, they are not part of an entry. Cookies are not supposed to be created by a client. They are only supposed to be retrieved through the log's `current_read_cookie()` and `current_append_cookie()` functions, which return the cookies identifying the next log entry to be read/appended, respectively. The client can hold on to these cookies and seek to them in the future.

The only exception to this is the special `SeekBeginning` enum, which represents the oldest remaining entry of the log. `SeekBeginning` is never returned by the `current_*_cookie()` functions, it is only used by the client to seek to the beginning of the log. Perhaps grouping it with the cookie under the `StorageCookie` enum isn't the best design. From these comments, I think that making the cookie a private associated type within the log and replacing the use of `SeekBeginning` with a `oldest_entry_cookie()` or `log_beginning_cookie()` function may make more sense.

The need for the ability to seek to the oldest remaining entry within a log and the reason why cookies are used in place of simple offsets is that entries can be overwritten in circular logs (see `capsules/src/log.rs`, this functionality is not mandated or discussed in the hil). When the underlying storage volume of a circular log is filled up, new entries will overwrite the oldest ones. This causes the beginning of the log to change as the oldest entries get overwritten, hence the need for the special `SeekBeginning` functionality. There is also need to distinguish between an old entry that was overwritten and the new entry in the same position that overwrote it, which is internally fulfilled by the cookie (which grants the newer entry a larger cookie value). If the log is given a cookie to an old, overwritten entry that no longer exists, then it will return an error that alerts the client of this. However, cookies are guaranteed to be valid as long as no entries are overwritten (if any entries are overwritten, then the client will be notified through the `records_lost` argument in the `append_done` callback, which is `true` if any number of records were overwritten by the append operation).

Hopefully this helps clarify, and sorry for not being clear before. I'll update the documentation with more detail once everything is clarified.

TPackard commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,88 @@
+//! Interface for a persistent log that stores distinct log entries.
+//!
+//! Log entries are appended to the end of a log and read back sequentially. Logs should be
+//! designed in such a way that they persist across device reboots.
```
No, the log format is left up to the implementer, the hil only outlines the interface to a log. The plan is for logs to be read by the process console.

TPackard commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,88 @@
+    fn set_read_client(&'a self, read_client: &'a dyn LogReadClient);
+
+    /// Read the next entry from the log. The log advances to the next entry after a successful
+    /// read.
```
The operation fails if the buffer is too small, the exact max entry size is left up to the implementer. For example (from the log capsule implementation):

https://github.com/tock/tock/blob/866d3f36241df00b2bc364a245da0b474467b010/capsules/src/log.rs#L592

TPackard commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,88 @@
+    /// Get the cookie representing the next entry to be read.
+    fn current_read_cookie(&self) -> LogCookie;
+
+    /// Seek to a new entry to read. The `LogCookie` for the entry should either be
```
It only affects the read behavior, I can change the name if that makes more sense.

TPackard commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,88 @@
+    /// until it is synced.
+    fn sync(&self) -> ReturnCode;
+
+    /// Erase the entire log.
```
This is left up to the implementer, but it zeros the data in the log capsule implementation.

https://github.com/tock/tock/blob/866d3f36241df00b2bc364a245da0b474467b010/capsules/src/log.rs#L788

TPackard commented:
Changed cookies to entry IDs, as the purpose of cookies was to identify different entries. Entry IDs are now an associated type within the `LogRead` trait instead of a public enum. The `log_start()`, `log_end()`, and `next_read_entry_id()` functions are now the ways to get different entries to seek to.

bradjc commented on kernel/src/hil/log.rs:
```diff
@@ -0,0 +1,92 @@
+
+    /// Read the next entry from the log. The log advances to the next entry after a successful
+    /// read. State does not change in the event of a failure.
+    fn read(
```
What does this return if there are no more entries to read?

phil-levis commented:
bors r+

TPackard commented:
> I have a question about the read() call, but otherwise looks good and I think the interface is more intuitive now.
> 
> Is there a corresponding tool to read back the log for debugging purposes? Is this something we should integrate into tockloader?

`read()` returns `ReturnCode::FAIL` if there are no more entries. I realized that I only put the return codes used in the implementation file, I'll document the important ones in the hil file too for the next PR.

The next PR is to use a log to store panic messages in the kernel (without halting the kernel) and allow the process console to read back this log.
